### DISCUSSION TITLE
How do I configure polling?

### DISCUSSION THREAD

**Discussion by LordMike** [zwave-js/zwave-js-ui#293] [Q&A]
Back in Z2m, I had a polling setup where I polled one specific device's value. It comes online, but doesn't report any values, so I need to poll it .. :/

I managed to set up polling previously, using the gateway settings where I by some stroke of luck managed to click just enough buttons for something to work.

I can see the same settings in zwavejs2mqtt - but I'm not sure what to put where. The docs are a little lacking here :)

### Sidenote
The dropdown with devices in the new-value dialog, does not include my device after a restart.. I thought it just needed to do its initial setup stuff to get all devices pinged, but even when other devices are in, this one isn't. 

Current devices.

![image](https://user-images.githubusercontent.com/1027111/105247001-609a7000-5b74-11eb-9859-a7a98082b634.png)

NorthQ Q-Gas 9121 has been "identified", and is in `RestartFromCache` state.

![image](https://user-images.githubusercontent.com/1027111/105247048-714ae600-5b74-11eb-86e1-33da0a6a8325.png)

Including node.json, for posterity [node_4.json.txt](https://github.com/zwave-js/zwavejs2mqtt/files/5845577/node_4.json.txt)

**Comment by LordMike** [comment]
From the fields in the new value dialog, it seems like it's not the same as it was in z2m? :)

**Comment by robertsLando** [comment]
@LordMike Actually zwavejs doesn't support polling. Polling is not easy and @AlCalzone is speaking with a zwave expert to understand the best way to implement it

**Reply by LordMike** [reply]
What if I were to send a command to get the value from the (sleeping) device.. it would be delivered when it wakes up (= polling), right?..

If I send two commands - are two delivered, or just the one ?

**Comment by cbho** [comment]
I'm aware that this might not be easy, but I really think this is an important feature.

Especially som older devices would benefit from this, e.g. the Greenwave Powernodes:
The Powernode 6 has 7 endpoints (root and six plugs) - and supports (in theory, that is, depending on (non-upgradeable) firmware version) status updates based on changes in power consumption, but is otherwise pretty "silent" unless polled.
However, if polled "as one device" the network is flooded when all walues from all endpoints are updated (I have five of these and on another system an error in the definition ended up polling them all every 1 (one) second - that was fun...)

Being able to configure polling on individual endpoints (and with individual intervals) would be sooo cool.

@robertsLando & @AlCalzone:
I don't know if this could be usefull at all, but [Homey](https://homey.app/da-dk/) is build on NodeJS and a lot of apps are are available on GitHub (e.g. one for the [Powernodes](https://github.com/TedTolboom/com.greenwavesystems/blob/PN5-only/drivers/powernode-6/driver.js)) - and a lot of work has been done by community developers on implementing a lot of devices (including improving the implementation of individual devices).
Their [developer guidelines](https://apps-sdk-v3.developer.athom.com/index.html) are also available  - perhaps some inspiration could be found there?

Unfortunately, I'm not proficient enough to help on the actual coding - but I'd be happy to help searching through stuff and go all-in on testing ;-)

Regards,
Chr.

**Reply by robertsLando** [reply]
https://github.com/zwave-js/node-zwave-js/issues/1163

**Reply by AlCalzone** [reply]
@cbho What you linked there seems like device-specific handlers. I would argue that is out of the scope of zwave-js (since there are thousands of devices) and should probably only be done to work around device-specific glitches.

I believe the recommended approach is to handle device types, like the HA team seems to be doing right now.

An API to handle polling will make it into zwave-js in the near future.

**Reply by LordMike** [reply]
If you made a `refreshValues` that only works on one (or more) select CC's / indexes, and ensure that no matter how many times the API is called, only one message is sent to a device when it wakes up (ie. the queue should deduplicate it).. Then it could be up to the user (zwavejs2mqtt) to have loops and timers to call this.

End users, like myself, could make our own timers and call the API through MQTT (or js), and that way around have "polling".. :)

See also #309

**Reply by AlCalzone** [reply]
Already started working on that 😎

**Reply by LordMike** [reply]
You're kinda putting my sponsorship to shame.

**Reply by AlCalzone** [reply]
https://github.com/zwave-js/node-zwave-js/pull/1479 is waiting for a test (you'll have to wait for @robertsLando to add support though)

**Reply by LordMike** [reply]
With that (and an API in mqtt). I would wager that the polling question is 98% handled. Users wanting it can make their own timers and do polling that way around.

**Reply by robertsLando** [reply]
Will work on that today

**Reply by robertsLando** [reply]
Pr ready #343
